[CSS Tree Counting Functions] Invalidation

weinig opened the pull request:
#### bc4b0a6362a439aac78cb889db08d6a1a862310b
<pre>
[CSS Tree Counting Functions] Invalidation
<a href="https://bugs.webkit.org/show_bug.cgi?id=291329">https://bugs.webkit.org/show_bug.cgi?id=291329</a>
<a href="https://rdar.apple.com/149410107">rdar://149410107</a>

Reviewed by Ryosuke Niwa.

Ensures that that MatchedDeclarationsCache is disabled and sibling changes
invalidate the style when using CSS sibling tree counting functions().

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-in-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/sibling-function-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/tree-scoped-sibling-function-expected.txt:
    - Update tests and results.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
    - Adds flag to RenderStyle for when using tree counting functions
      and disable the MatchedDeclarationsCache when it is set.

* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
    - Set the new RenderStyle flag for tree counting functions and set
      both ChildrenAffectedBy{Backward,Forward}PositionalRules flags
      on the parent element to ensure sibling changes invalidate the
      element&apos;s style. A more specific flag could be used in the future
      that only invalidates if the appropriate count/index changes, but
      that would require adding addition bits to m_styleBitfields which
      is already at capacity.

Canonical link: <a href="https://commits.webkit.org/294062@main">https://commits.webkit.org/294062@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb1a9850c446a25036c29ed69d7ff1777bed3eee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100586 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105723 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51174 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102627 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20546 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28712 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76602 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33638 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103593 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15766 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90861 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56958 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15625 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8866 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50550 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85495 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108077 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27704 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20348 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85550 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28067 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87061 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85088 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21680 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29790 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7515 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21692 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27639 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32890 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27450 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30768 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29008 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,29 +333,69 @@ double BuilderState::lookupCSSRandomBaseValue(const CSSCalc::RandomCachingKey& k
-        ++count;
-    for (const auto* sibling = ElementTraversal::nextSibling(*element()); sibling; sibling = ElementTraversal::nextSibling(*sibling))
-        ++count;
+    // FIXME: This is not exactly what we need. We something that will return nullptr if the parent would have been in an outer shadow scope. In that case, we should return 0;
```
You probably want to take in account `state.styleScopeOrdinal()` in some way here. The other usages should give you an idea how it can be used.

weinig replied:
I'm honestly not sure how to do this even after looking at the other uses. 

I think I need to compare two different style scope ordinals. The one for the current `element()` I guess is `state.styleScopeOrdinal()`. But then I also need one for `element()->parentElement()`, which I am not clear on how I would get that.

weinig replied:
Although, maybe this is no longer relevant now that it is the DOM tree instead of the flat tree being used.

nt1m commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,29 +333,69 @@ double BuilderState::lookupCSSRandomBaseValue(const CSSCalc::RandomCachingKey& k
 
     ASSERT(element());
 
+    // FIXME: This is not exactly what we need. We something that will return nullptr if the parent would have been in an outer shadow scope. In that case, we should return 0;
```
ditto

rniwa commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,29 +333,69 @@ double BuilderState::lookupCSSRandomBaseValue(const CSSCalc::RandomCachingKey& k
-    for (const auto* sibling = ElementTraversal::nextSibling(*element()); sibling; sibling = ElementTraversal::nextSibling(*sibling))
-        ++count;
+    // FIXME: This is not exactly what we need. We something that will return nullptr if the parent would have been in an outer shadow scope. In that case, we should return 0;
+    RefPtr parent = element()->parentElementInComposedTree();
```
Oh, I don't think this is what we wanna use based on the spec text. We probably want to use `parentElement()` instead.

weinig replied:
Which spec test specifically? The note in the spec says: “Like the rest of CSS (other than [selectors](https://www.w3.org/TR/selectors-4/#selector)), [sibling-count()](https://drafts.csswg.org/css-values-5/#funcdef-sibling-count) and [sibling-index()](https://drafts.csswg.org/css-values-5/#funcdef-sibling-index) operate on the [flat tree](https://drafts.csswg.org/css-scoping-1/#flat-tree).” 

I thought that’s what this does. 

It also says “When used on an [element-backed pseudo-element](https://drafts.csswg.org/css-pseudo-4/#element-backed) which is also a real element, the tree counting functions resolve for that real element. For other pseudo elements, they resolve as if they were resolved against the originating element. It follows that for nested pseudo elements the resolution will recursively walk the originating elements until a real element is found.” 

Though I am less clear on that part so far.

rniwa replied:
oh weird. Example 30 contradicts that statement: https://drafts.csswg.org/css-values-5/#example-b28bb265

rniwa replied:
@fantasai : are these functions supposed to work on the original DOM tree or the composed/flat tree?

weinig replied:
There are so many examples in example 30 😀. Which one do you mean?

rniwa replied:
```
::slotted(*)::before {
  /* Based on the sibling-index() of the slotted element - outer tree */
  width: calc(sibling-index() * 10px);
}
```
For this one, the comment says we should be using sibling-index of the slotted element in the outer tree. That implies that we use `nodeIndex` of the assigned node, not the index within the assigned nodes.

rniwa commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,29 +333,69 @@ double BuilderState::lookupCSSRandomBaseValue(const CSSCalc::RandomCachingKey& k
+    parent->setChildrenAffectedByForwardPositionalRules();
+
+    unsigned count = 0;
+    for (auto& node : composedTreeChildren(*parent)) {
```
Ditto. Based on the spec text, we probably want to use regular children, not children in the composed tree.

weinig replied:
I think this (or something other than the normal tree) was needed to fix https://github.com/WebKit/WebKit/pull/43855/files#diff-ac140730e846a8c05e487ce055b85cc59997269fbd2ced4fea5ee532f12f54b8

rniwa commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,29 +333,69 @@ double BuilderState::lookupCSSRandomBaseValue(const CSSCalc::RandomCachingKey& k
     ASSERT(element());
 
+    // FIXME: This is not exactly what we need. We something that will return nullptr if the parent would have been in an outer shadow scope. In that case, we should return 0;
+    RefPtr parent = element()->parentElementInComposedTree();
```
Ditto.

rniwa commented on Source/WebCore/style/StyleBuilderState.cpp:
```diff
@@ -332,29 +333,69 @@ double BuilderState::lookupCSSRandomBaseValue(const CSSCalc::RandomCachingKey& k
     unsigned count = 1;
-    for (const auto* sibling = ElementTraversal::previousSibling(*element()); sibling; sibling = ElementTraversal::previousSibling(*sibling))
-        ++count;
+    for (auto& node : composedTreeChildren(*parent)) {
```
Ditto.

weinig committed:
[CSS Tree Counting Functions] Invalidation
https://bugs.webkit.org/show_bug.cgi?id=291329
rdar://149410107

Reviewed by Ryosuke Niwa.

Ensures that that MatchedDeclarationsCache is disabled and sibling changes
invalidate the style when using CSS sibling tree counting functions().

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function-in-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/calc-sibling-function.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/sibling-function-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/tree-counting/tree-scoped-sibling-function-expected.txt:
    - Update tests and results.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
    - Adds flag to RenderStyle for when using tree counting functions
      and disable the MatchedDeclarationsCache when it is set.

* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
    - Set the new RenderStyle flag for tree counting functions and set
      both ChildrenAffectedBy{Backward,Forward}PositionalRules flags
      on the parent element to ensure sibling changes invalidate the
      element's style. A more specific flag could be used in the future
      that only invalidates if the appropriate count/index changes, but
      that would require adding addition bits to m_styleBitfields which
      is already at capacity.

Canonical link: https://commits.webkit.org/294062@main
